src/graph/graph.c

```diff
@@ -34,14 +34,31 @@
     g->nodes_blocks[GRAPH_NODE_ID_TO_BLOCK_INDEX(id)]
 
 /*========================= Graph utility functions ========================= */
+
+/* Acquire mutex. */
+void _Graph_EnterCriticalSection(Graph *g) {
+    pthread_mutex_lock(&g->_mutex);
+}
+
+/* Release mutex. */
+void _Graph_LeaveCriticalSection(Graph *g) {
+    pthread_mutex_unlock(&g->_mutex);
+}
+
 // Resize given matrix to match graph's adjacency matrix dimensions.
 void _Graph_ResizeMatrix(const Graph *g, GrB_Matrix m) {
     GrB_Index n_rows;
-    
+
     GrB_Matrix_nrows(&n_rows, m);
     if(n_rows != g->node_count) {
-        GrB_Info res = GxB_Matrix_resize(m, g->node_count, g->node_count);
-        assert(res == GrB_SUCCESS);
+        _Graph_EnterCriticalSection((Graph *)g);
+        {
+            // Double check now that we're in critical section.
+            GrB_Matrix_nrows(&n_rows, m);
+            if(n_rows != g->node_count)
+                assert(GxB_Matrix_resize(m, g->node_count, g->node_count) == GrB_SUCCESS);
+        }
+        _Graph_LeaveCriticalSection((Graph *)g);
```

Comment:
You're right, this is could be a future optimisation in case we see concurrent read threads fighting over this mutex when trying to work with different matrices, I'll add a comment.